### PR TITLE
RavenDB-21679 Restrict the import of features with user certificates

### PR DESCRIPTION
**Pull request by efratshenhar**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-21679

### Additional description

When performing an import with a client certificate (read/write), we failed to enforce feature restrictions.
The import process will now skip the following restricted features: 
- Indexes ( not JavaScript)
- Connection strings
- Expiration
- Periodic backups
- Client configuration
- Revision Configuration

_Please delete below the options that are not relevant_

### Type of change

- Bug fix

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.

### CODE REVIEW

File: src/Raven.Server/Smuggler/Documents/Actions/DatabaseRecordActions.cs [hunk 1]
```diff
@@ -88,7 +88,7 @@ public DatabaseRecordActions(ServerStore server, DatabaseRecord currentDatabaseR
             result.DatabaseRecord.ConflictSolverConfigUpdated = true;
         }
 
-        if (databaseRecord.PeriodicBackups.Count > 0 && databaseRecordItemType.HasFlag(DatabaseRecordItemType.PeriodicBackups))
+        if (CanAccess(authorizationStatus) && databaseRecord.PeriodicBackups.Count > 0 && databaseRecordItemType.HasFlag(DatabaseRecordItemType.PeriodicBackups))
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/Actions/DatabaseRecordActions.cs]: We need to alert on those , no?

File: src/Raven.Server/Smuggler/Documents/SmugglerBase.cs [hunk 2]
```diff
@@ -713,6 +714,13 @@ await foreach (var index in _source.GetIndexesAsync())
                         continue;
                     }
 
+                    if (_options.AuthorizationStatus == AuthorizationStatus.ValidUser &&
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/SmugglerBase.cs]: This will filter things out silently, which can be _super_ confusing to the user.

Comment (grisha-kotler) [reply src/Raven.Server/Smuggler/Documents/SmugglerBase.cs]: We're adding an error to the import log.

Comment (ayende) [reply src/Raven.Server/Smuggler/Documents/SmugglerBase.cs]: In the UI, is that clearly visible?

File: src/Raven.Server/Smuggler/Documents/Actions/DatabaseRecordActions.cs (hunk 1 above)
Comment (efratshenhar) [reply src/Raven.Server/Smuggler/Documents/Actions/DatabaseRecordActions.cs]: image

File: src/Raven.Server/Smuggler/Documents/SmugglerBase.cs (hunk 2 above)
Comment (efratshenhar) [reply src/Raven.Server/Smuggler/Documents/SmugglerBase.cs]: image

**Comment by ppekrol** [comment]
Status?

**Comment by arekpalinski** [comment]
@grisha-kotler please make sure it's reviewed by you once you're back

**Comment by arekpalinski** [comment]
What's the status here?

**Review by grisha-kotler — APPROVED** [review]

File: test/SlowTests/RavenDB-21679.cs [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+                using (context.OpenReadTransaction())
+                {
+                    record = Server.ServerStore.Cluster.ReadDatabase(context, store.Database);
+                }
+
+                Assert.False(record.RavenConnectionStrings.ContainsKey("RavenConnectionString"));
+                Assert.Null(record.Expiration);
+                Assert.False(record.PeriodicBackups.Count > 0);
+                Assert.Null(record.Client);
+                Assert.Null(record.Revisions);
```
Comment (arekpalinski) [inline test/SlowTests/RavenDB-21679.cs]: Can we also verify that the relevant warnings / errors have been written to the import log?

File: src/Raven.Server/Documents/Handlers/Processors/SampleData/SampleDataHandlerProcessorForPostSampleData.cs [hunk 4]
```diff
@@ -17,17 +20,26 @@ public SampleDataHandlerProcessorForPostSampleData([NotNull] DatabaseRequestHand
...
-                var smuggler = RequestHandler.Database.Smuggler.Create(source, destination, context,
-                    options: new DatabaseSmugglerOptionsServerSide
-                    {
-                        OperateOnTypes = operateOnTypes,
-                        SkipRevisionCreation = true
-                    });
-
+                };
+                options.AuthorizationStatus = AuthorizationStatus.ValidUser;
+                if (feature != null )
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Handlers/Processors/SampleData/SampleDataHandlerProcessorForPostSampleData.cs:35]: Below (`ShardedSampleDataHandlerProcessorForPostSampleData.cs`) we're checking if auth feature is avaialble, if not then we set `options.AuthorizationStatus = AuthorizationStatus.ValidUser;`

https://github.com/ravendb/ravendb/pull/17907/files#diff-88b758d616a14aafedd2cbf29655b3b57e36fbdad05dc8f95468ad86bba7e185R33-R34

Don't we need the same here?

File: src/Raven.Server/Smuggler/Documents/Actions/DatabaseIndexActions.cs [hunk 5]
```diff
@@ -26,8 +30,15 @@ public DatabaseIndexActions([NotNull] AbstractIndexCreateController controller,
             _batch = _controller.CreateIndexBatch();
     }
 
-    public async ValueTask WriteIndexAsync(IndexDefinitionBaseServerSide indexDefinition, IndexType indexType)
+    public async ValueTask WriteIndexAsync(IndexDefinitionBaseServerSide indexDefinition, IndexType indexType, AuthorizationStatus authorizationStatus)
     {
+        if (_configuration.Security.AuthenticationEnabled &&
+            authorizationStatus == AuthorizationStatus.ValidUser)
+        {
+            throw new CertificateException(
```
Comment (arekpalinski) [inline src/Raven.Server/Smuggler/Documents/Actions/DatabaseIndexActions.cs:38]: Don't we need to distinguish between C# indexes and JS indexes?

File: src/Raven.Server/Smuggler/Documents/Actions/DatabaseIndexActions.cs [hunk 6]
```diff
@@ -38,8 +49,17 @@ public async ValueTask WriteIndexAsync(IndexDefinitionBaseServerSide indexDefini
         await _controller.CreateIndexAsync(indexDefinition, RaftIdGenerator.DontCareId);
     }
 
-    public async ValueTask WriteIndexAsync(IndexDefinition indexDefinition)
+    public async ValueTask WriteIndexAsync(IndexDefinition indexDefinition, AuthorizationStatus authorizationStatus)
     {
+        if (_configuration.Security.AuthenticationEnabled &&
+            authorizationStatus == AuthorizationStatus.ValidUser &&
+            (indexDefinition.Type == IndexType.Map ||
+             indexDefinition.Type == IndexType.MapReduce))
```
Comment (arekpalinski) [inline src/Raven.Server/Smuggler/Documents/Actions/DatabaseIndexActions.cs:57]: Why only if this is Map or MapReduce?

Comment (arekpalinski) [reply src/Raven.Server/Smuggler/Documents/Actions/DatabaseIndexActions.cs:57]: We don't have the checks about Type in the above method:
https://github.com/ravendb/ravendb/pull/17907/files#diff-bfeecfcaf9cd5a808953831746e8f93b98d1470dd355aff6de46ee6737c66c68R35-R36

File: src/Raven.Server/Smuggler/Documents/Actions/DatabaseRecordActions.cs [hunk 7]
```diff
@@ -607,6 +645,13 @@ public DatabaseRecordActions(ServerStore server, DatabaseRecord currentDatabaseR
         tasks.Clear();
     }
 
+    private static bool CanAccess(AuthorizationStatus authorizationStatus)
```
Comment (arekpalinski) [inline src/Raven.Server/Smuggler/Documents/Actions/DatabaseRecordActions.cs:649]: What is this is a read only access? We have the notion of read only cerificates. Do we check that as well?

File: src/Raven.Server/Smuggler/Documents/Actions/DatabaseIndexActions.cs (hunk 5 above)
Comment (efratshenhar) [reply src/Raven.Server/Smuggler/Documents/Actions/DatabaseIndexActions.cs:38]: We are getting here only with Auto indexes

File: src/Raven.Server/Smuggler/Documents/Actions/DatabaseIndexActions.cs (hunk 6 above)
Comment (efratshenhar) [reply src/Raven.Server/Smuggler/Documents/Actions/DatabaseIndexActions.cs:57]: Here we can get with IndexType.Map ,IndexType.MapReduce and JS index and we don't want to throw for JS indexes

File: src/Raven.Server/Smuggler/Documents/Actions/DatabaseRecordActions.cs (hunk 7 above)
Comment (efratshenhar) [reply src/Raven.Server/Smuggler/Documents/Actions/DatabaseRecordActions.cs:649]: read-only access is checked at the endpoint level

File: src/Raven.Server/Documents/Handlers/Processors/SampleData/SampleDataHandlerProcessorForPostSampleData.cs (hunk 4 above)
Comment (efratshenhar) [reply src/Raven.Server/Documents/Handlers/Processors/SampleData/SampleDataHandlerProcessorForPostSampleData.cs:35]: we are setting options.AuthorizationStatus = AuthorizationStatus.ValidUser before we check if feature != null. so if feature is null we don't change AuthorizationStatus .I will change in Sharded for consistency

File: src/Raven.Server/Smuggler/Documents/Actions/DatabaseIndexActions.cs (hunk 5 above)
Comment (arekpalinski) [reply src/Raven.Server/Smuggler/Documents/Actions/DatabaseIndexActions.cs:38]: Can we rename the method to `WriteAutoIndexAsync` then?

File: src/Raven.Server/Smuggler/Documents/Actions/DatabaseIndexActions.cs (hunk 6 above)
Comment (arekpalinski) [reply src/Raven.Server/Smuggler/Documents/Actions/DatabaseIndexActions.cs:57]: Okay. Another question - you're checking `authorizationStatus == AuthorizationStatus.ValidUser`, can `authorizationStatus` be `RestrictedAccess`?

File: src/Raven.Server/Smuggler/Documents/Actions/DatabaseRecordActions.cs (hunk 7 above)
Comment (arekpalinski) [reply src/Raven.Server/Smuggler/Documents/Actions/DatabaseRecordActions.cs:649]: Okay. I have additional question then. In the import endpoint we have `AuthorizationStatus.DatabaseAdmin` specified:

https://github.com/ravendb/ravendb/blob/8c28e427978cf50b98450d482a049d3ba1592f06/src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs#L63-L64

So how did we manage to do the import with `ValidUser` access?

Comment (efratshenhar) [reply src/Raven.Server/Smuggler/Documents/Actions/DatabaseRecordActions.cs:649]: If we use smuggler directly and won't go from the endpoint like in the test I added.

File: src/Raven.Server/Smuggler/Documents/Actions/DatabaseIndexActions.cs (hunk 6 above)
Comment (efratshenhar) [reply src/Raven.Server/Smuggler/Documents/Actions/DatabaseIndexActions.cs:57]: No. RestrictedAccess can be set only when we use the endpoint that will block the import completely. ValidUser can be set bt the Certificate type and when we use smuggler directly

File: src/Raven.Server/Smuggler/Documents/Actions/DatabaseIndexActions.cs (hunk 5 above)
Comment (efratshenhar) [reply src/Raven.Server/Smuggler/Documents/Actions/DatabaseIndexActions.cs:38]: Done

File: test/SlowTests/RavenDB-21679.cs (hunk 3 above)
Comment (efratshenhar) [reply test/SlowTests/RavenDB-21679.cs]: Done

File: src/Raven.Server/Smuggler/Documents/Actions/DatabaseRecordActions.cs (hunk 7 above)
Comment (arekpalinski) [reply src/Raven.Server/Smuggler/Documents/Actions/DatabaseRecordActions.cs:649]: Right, now it's clear.

File: src/Raven.Server/Smuggler/Documents/Actions/DatabaseIndexActions.cs (hunk 6 above)
Comment (arekpalinski) [reply src/Raven.Server/Smuggler/Documents/Actions/DatabaseIndexActions.cs:57]: OK

**Review by arekpalinski — APPROVED** [review]

**Pull request merged**